presto-iceberg/src/test/java/com/facebook/presto/iceberg/procedure/TestExpireSnapshotProcedure.java

```diff
@@ -0,0 +1,289 @@
+            Table table = loadTable(tableName);
+            long snapshotId1 = table.currentSnapshot().snapshotId();
+            assertUpdate(session, "INSERT INTO " + tableName + " VALUES(2, 'b')", 1);
+            assertUpdate(session, "INSERT INTO " + tableName + " VALUES(3, 'c')", 1);
+            table.refresh();
+            long snapshotId2 = table.currentSnapshot().snapshotId();
+            assertUpdate(session, "INSERT INTO " + tableName + " VALUES(4, 'd')", 1);
+
+            table.refresh();
+            assertHasSize(table.snapshots(), 4);
+
+            // Drop the indicated 2 snapshots
+            assertUpdate(session, format("CALL system.expire_snapshots(table_name => '%s', schema => '%s', snapshot_ids => %s)",
+                    tableName, TEST_SCHEMA, "array[" + snapshotId1 + ", " + snapshotId2 + "]"));
+            table.refresh();
+            assertHasSize(table.snapshots(), 2);
+        }
+        finally {
+            dropTable(tableName);
+        }
+    }
+
+    @Test
+    public void testInvalidExpireSnapshotsCases()
+    {
+        assertQueryFails("CALL system.expire_snapshots('n', table_name => 't')", ".*Named and positional arguments cannot be mixed");
+        assertQueryFails("CALL custom.expire_snapshots('n', 't')", "Procedure not registered: custom.expire_snapshots");
+        assertQueryFails("CALL system.expire_snapshots()", ".*Required procedure argument 'schema' is missing");
+        assertQueryFails("CALL system.expire_snapshots('s', 'n', 2.2)", ".*Cannot cast type decimal\\(2,1\\) to timestamp");
+        assertQueryFails("CALL system.expire_snapshots('', '')", "schemaName is empty");
+    }
+
+    private String getTimestampString(long timeMillsUtc, String zoneId)
+    {
+        Instant instant = Instant.ofEpochMilli(timeMillsUtc);
+        LocalDateTime localDateTime = instant
+                .atZone(ZoneId.of(zoneId))
+                .toLocalDateTime();
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss.SSS");
+        formatter = formatter.withZone(ZoneId.of(zoneId));
+        return localDateTime.format(formatter);
+    }
+
+    protected Table loadTable(String tableName)
+    {
+        Catalog catalog = CatalogUtil.loadCatalog(HadoopCatalog.class.getName(), ICEBERG_CATALOG, getProperties(), new Configuration());
+        return catalog.loadTable(TableIdentifier.of(TEST_SCHEMA, tableName));
+    }
+
+    protected Map<String, String> getProperties()
+    {
+        File metastoreDir = getCatalogDirectory();
+        return ImmutableMap.of("warehouse", metastoreDir.toString());
+    }
+
+    protected File getCatalogDirectory()
+    {
+        Path dataDirectory = getDistributedQueryRunner().getCoordinator().getDataDirectory();
+        return dataDirectory.toFile();
+    }
+
+    protected long waitUntilAfter(long timestampMillis)
+    {
+        long current = System.currentTimeMillis();
```

Comment:
This is run in a hot loop.  I don't see where we actually wait though--this seems to always wait until after the timestamp of the snapshot.  Do we need it?

Reply:
I think we need this. Iceberg's `ExpireSnapshots.expireOlderThan(timestamp)` will expires all snapshots older than the given milliseconds timestamp, that means the snapshot with the same timestamp would not be expired. So for the sake of stability in testing, we'd better wait for this millisecond to pass, in case that all things happen in the same millisecond.

Reply:
Can we not just use `Thread.sleep(1)` in place of this?

Reply:
Oh, I referred to Iceberg's test cases which use this spinning wait. In my opinion, when the wait time is very short, for example shorter than 1ms in this case, a spinning wait may be better than a block wait. It could avoid thread scheduling and context switching which may be triggered by `Thread.sleep(ts)`, that might cause the waiting time more longer than expected.

Besides, should we obey the test standards to avoid using `Thread.sleep`? Referring to https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md?plain=1#L374

Reply:
If it's shorter than 1ms, let's add a precondition which asserts that, so that the test fails rather than busy waits?

Reply:
Sure, that makes sense! Fixed.